job_controller/database/db_updater.py

```diff
@@ -0,0 +1,276 @@
+            filename=filename,
+            title=title,
+            users=users,
+            experiment_number=experiment_number,
+            run_start=run_start,
+            run_end=run_end,
+            good_frames=good_frames,
+            raw_frames=raw_frames,
+        )
+        reduction = Reduction(
+            reduction_start=None,
+            reduction_end=None,
+            reduction_state=None,
+            reduction_inputs=reduction_inputs,
+            script=None,
+            reduction_outputs=None,
+        )
+        session.add(run)
+        session.add(reduction)
+        session.commit()
+        # Now create the run_reduction entry and add it
+        run_reduction = RunReduction(run=run.id, reduction=reduction.id)
+        session.add(run_reduction)
+        session.commit()
+
+        logger.info(
+            "Submitted detected-run to the database successfully: {filename: %s, title: %s, users: %s, "
+            "experiment_number: %s, run_start: %s, run_end: %s, good_frames: %s, raw_frames: %s, "
+            "reduction_inputs: %s}",
+            filename,
+            title,
+            users,
+            experiment_number,
+            run_start,
+            run_end,
+            good_frames,
+            raw_frames,
+            reduction_inputs,
+        )
+
+        return reduction.id.one()
+
+    def add_completed_run(
+        self,
+        db_reduction_id: int,
+        reduction_inputs: Dict[str, Any],
+        state: State,
+        status_message: str,
+        output_files: List[str],
+        reduction_script: str,
+    ) -> None:
+        """
+        This function submits data to the database from what is initially available on completed-runs kafka topic
+        :param db_reduction_id: The ID for the reduction row in the reduction table
+        :param reduction_inputs: The inputs used in the reduction script by the IR-API
+        :param state: The state of how the run ended
+        :param status_message: The message that accompanies the state for how the  state ended, if the state for
+        example was unsuccessful or an error, it would have the reason/error message.
+        :param output_files: The files output from the reduction job
+        :param reduction_script: The script used in the reduction
+        :return:
+        """
+        logger.info(
+            "Submitting completed-run to the database: {id: %s, reduction_inputs: %s, state: %s, "
+            "status_message: %s, output_files: %s, reduction_script: %s}",
+            db_reduction_id,
+            reduction_inputs,
+            str(state),
+            status_message,
+            output_files,
+            reduction_script,
+        )
+        session = self.session_maker_func()
+        script = session.query(Script).filter_by(script=reduction_script).first()
+        if script is None:
+            script = Script(script=reduction_script)
+            session.add(script)
+            session.commit()
```

Comment:
Same comment as above re atomicity.

Reply:
Script.id needs to be generated by the PostgreSQL db